crates/rust-analyzer/src/handlers.rs

```diff
@@ -273,19 +273,24 @@ pub(crate) fn handle_document_symbol(
         parents.push((doc_symbol, symbol.parent));
     }
     let mut document_symbols = Vec::new();
+    // Constructs `document_symbols` from `parents`, in order from the end.
     while let Some((node, parent)) = parents.pop() {
         match parent {
             None => document_symbols.push(node),
```

Comment:
```suggestion
            None => {
                if let Some(children) = &mut node.children {
                    children.reverse()
                }
                document_symbols.push(node)
            },
```

I think should do the trick without recursion.

Reply:
Thanks for the suggestion.
I tried your code, but it cannot deal with nested symbols.

```rust
mod a {
    mod b {
        struct B1;
        struct B2;
    }
}
```

We want this snippet to be output like:

```json
[
  {
    "children": [
      {
        "children": [
          {
            "kind": 23,
            "name": "B1"
          },
          {
            "kind": 23,
            "name": "B2"
          }
        ],
        "kind": 2,
        "name": "b"
      }
    ],
    "kind": 2,
    "name": "a"
  }
]
```

But actually we get:

```json
[
  {
    "children": [
      {
        "children": [
          {
            "kind": 23,
            "name": "B2"
          },
          {
            "kind": 23,
            "name": "B1"
          }
        ],
        "kind": 2,
        "name": "b"
      }
    ],
    "kind": 2,
    "name": "a"
  }
]
```

Recursion way works as we expect.